### ISSUE TITLE
convert.py - RuntimeError: CUDA error: invalid configuration argument

### ISSUE THREAD

**Issue by Thireus** [turboderp/exllamav2#20]
```
python convert.py -i ~/safetensor/model -o ~/EXL2/model_4bit -c ~/EXL2/0000.parquet -b 4.0 -hb 6 -l 4096 -ml 4096

 -- Beginning new job
 -- Input: /home/user/safetensor/model
 -- Output: /home/user/EXL2/model_4bit
 -- Calibration dataset: /home/user/EXL2/0000.parquet, 100 / 16 (16) rows, 4096 tokens per sample
 -- Target bits per weight: 4.0 (decoder), 6 (head)
 -- Tokenizing samples (measurement)...
/home/user/exllamav2/conversion/tokenize.py:16: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  tokens = tokenizer.encode(row[0])
 -- Token embeddings (measurement)...
 -- Measuring quantization impact...
 -- Layer: model.layers.0 (Attention)
Traceback (most recent call last):
  File "/home/user/exllamav2/convert.py", line 168, in <module>
    measure_quant(job, save_job, model)
  File "/home/user/exllamav2/conversion/quantize.py", line 184, in measure_quant
    outputs = module.forward(x, cache, attn_mask, intermediates = True)
  File "/home/user/exllamav2/exllamav2/attn.py", line 195, in forward
    return self.forward_torch(hidden_states, cache, attn_mask, past_len, intermediates)
  File "/home/user/exllamav2/exllamav2/attn.py", line 460, in forward_torch
    key_states = self.repeat_kv(key_states, self.model.config.num_key_value_groups)
  File "/home/user/exllamav2/exllamav2/attn.py", line 188, in repeat_kv
    hidden_states = hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
RuntimeError: CUDA error: invalid configuration argument
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

Any idea what could be the cause of this issue?

Edit: Bad or unsupported safetensor files it seems
Edit2: That wasn't bad safetensors in the end, cf TL;DR

---

TL;DR: The issue is likely due to VRAM limitation, 24GB not enough for -ml 4096 -l 4096 at the moment. Some VRAM otpimization code is underway, until then you must use the default -ml 2048 -l 2048.

**Comment by DocShotgun** [comment]
How exactly did you manage to resolve this? Getting this error right now.

I simply loaded my model and saved it in safetensors using transformers with model.save_pretrained() and safe_serialization = True. Not sure why the safetensor files would be bad after that.

**Comment by Thireus** [comment]
I haven't resolved it yet due to lack of time.

So far I've used https://github.com/oobabooga/text-generation-webui/blob/main/convert-to-safetensors.py which led to the issue and realized that it's a safetensors conversion problem.

I've asked around from people who successfully converted to good safetensors, and I'm now trying their custom scripts:
- https://huggingface.co/Panchovix/airoboros-l2-70b-gpt4-1.4.1-safetensors/blob/main/bin2safetensors/convert.py (https://huggingface.co/Panchovix/airoboros-l2-70b-gpt4-1.4.1_5.0bpw-h6-exl2/discussions/1#65023208a33ce3ae285b8924)
- https://gist.github.com/epicfilemcnulty/1f55fd96b08f8d4d6693293e37b4c55e (https://github.com/turboderp/exllamav2/issues/25#issuecomment-1718270070)

The scripts above should work. I'm currently trying in the middle of converting with epicfilemcnulty's script.

**Comment by turboderp** [comment]
There are a couple of issues currently with BF16 models that should be resolved shortly. This could also be a VRAM issue. Torch (especially with extensions that maybe don't do all the error handling they should be doing) can sometimes end up in an invalid state rather than crashing with an out-of-memory error, and due to the asynchronous nature of CUDA the function that triggers an exception is often not the one that's actually failing.

First, make sure you're not running out of memory. Run `nvtop` or some such in a separate terminal, check if you're running low on memory. Also, you don't have to calibrate on the full sequence length of the model. I find 2048 tokens works well enough even for a 4k-context model.

The implementation assumes that all tensors are contiguous when loaded from the .safetensors file(s), so that's something to be mindful of.

And while it should handle FP32 weights when loading an uncoverted model, they will use a lot more memory, so that's another possible source of OoM issues. Is it possible the difference between the "good" and "bad" .safetensors files in this case is that the good ones are already downcast to FP16?

**Comment by Thireus** [comment]
I see, thank you for the additional tips @turboderp, I'm also strongly guessing BF16-related issue - I'll have a look at VRAM usage, I'm using a 3090, so I believe VRAM shouldn't be the issue.

- Successful results with CodeLlama-13B-Instruct-fp16 - https://huggingface.co/TheBloke/CodeLlama-13B-Instruct-fp16
- Unsuccessful results when converting https://huggingface.co/WizardLM/WizardLM-70B-V1.0 to safetensors, also tried Float16 version from https://huggingface.co/simsim314/WizardLM-70B-V1.0-HF to safetensors.

**Comment by Thireus** [comment]
@turboderp, I can confirm this isn't a VRAM issue.

I have also tested your latest commit (5ccd29c7d639da7b33a16d97f3ad2df10ebab56a). I saw you made some changes to print errors when safetensor files don't have the expected data. But I'm still getting the same error message without further information.

I don't think it should matter, but I'm converting from .bin to .safetensor using a separate machine with cpu only.

**Comment by turboderp** [comment]
Are you also getting the deprecation warning at the start? I'm starting to think that may have something to do with it. There's another issue I'm working on where that also shows up, so it could have something to do with the calibration data being loaded incorrectly.

**Comment by Thireus** [comment]
@turboderp, I can confirm this isn't a VRAM issue.

I have also tested your latest commit (5ccd29c7d639da7b33a16d97f3ad2df10ebab56a). I saw you made some changes to print errors when safetensor files don't have the expected data. But I'm still getting the same error message without further information.

**Comment by DocShotgun** [comment]
Some additional details from my end:

1. I also encountered this error while attempting to quant a 70b model, while I previously quanted 13b models without issue.
2. The environment was an L4 runpod instance with the base torch 2.0.1 docker and python 3.10. It has 250gb ram and 48gb vram, which should be adequate for this.
3. The error occurs almost immediately when attempting to do measurement on attention layer 0.

**Comment by turboderp** [comment]
Do you get the same error without the `-l 4096 -ml 4096`. Cause that really shouldn't be needed, and while it shouldn't matter with 48 GB of VRAM, it does still use up memory.

Also, I'm wondering if the Parquet file you're using contains enough data for 409600 tokens. I know at least there's a wikitext-test split floating around with some 600 kB of text. If that runs out, I'm not sure what would happen.

**Comment by DocShotgun** [comment]
I ended up killing the instance before trying without the 4096 calibration context, so unsure on that one lol. 

But I used the same parquet with several megabytes of data that I used for the 13b quants at 4096 calibration, so there should be enough data there. It seemed to get past the tokenization of the dataset properly, as it’s after it says it’s going to work on attention layer 0 that I errored out.

**Comment by Thireus** [comment]
Happy to report that with your latest commit and without using "-ml 4096", the issue is gone.

```
python convert.py -i ~/safetensor/model -o ~/EXL2/model_4bit -c ~/EXL2/0000.parquet -b 4.0 -hb 6 -l 4096
 -- Beginning new job
 -- Input: /home/user/safetensor/model
 -- Output: /home/user/EXL2/model_4bit
 -- Calibration dataset: /home/user/EXL2/0000.parquet, 100 / 16 (16) rows, 4096 tokens per sample
 -- Target bits per weight: 4.0 (decoder), 6 (head)
 -- Tokenizing samples (measurement)...
 -- Token embeddings (measurement)...
 -- Measuring quantization impact...
 -- Layer: model.layers.0 (Attention)
 -- Linear: model.layers.0.self_attn.q_proj
 -- 0.05:3b/0.95:2b 32g s4         2.17 bpw    rfn_error: 0.00967
 -- 0.25:3b/0.75:2b 32g s4         2.38 bpw    rfn_error: 0.00922
 -- 0.1:4b/0.4:3b/0.5:2b 32g s4    2.72 bpw    rfn_error: 0.00573
 -- 0.1:4b/0.9:3b 32g s4           3.22 bpw    rfn_error: 0.00560
 -- 1.0:3b 128g s4                 3.03 bpw    rfn_error: 0.01691
 -- 1.0:3b 32g s4                  3.13 bpw    rfn_error: 0.00896
 -- 0.05:4b/0.95:3b 32g s4         3.17 bpw    rfn_error: 0.00566
```

But most importantly!!!! The output directory needs to be empty first. `rm -rf  ~/EXL2/model_4bit && mkdir ~/EXL2/model_4bit`, otherwise for some reason it will attempt to reuse "-ml 4096" again even though it's no longer mentioned as a parameter of the command. That behaviour made troubleshooting this issue quite hard.

**Comment by Thireus** [comment]
Well, after 22h of quantization, that happened:

```
 -- 0.4:6b/0.6:5b 32g s4           5.53 bpw    rfn_error: 0.01424
 -- 1.0:6b 128g s4                 6.03 bpw    rfn_error: 0.00961
 -- 0.1:8b/0.9:6b 128g s4          6.22 bpw    rfn_error: 0.00936
 -- 1.0:8b 32g s4                  8.13 bpw    rfn_error: 0.00747
 -- Time: 14.58 seconds
 -- Linear: model.layers.79.mlp.down_proj
 -- 0.05:3b/0.95:2b 32g s4         2.17 bpw    rfn_error: 0.04251
 -- 0.25:3b/0.75:2b 32g s4         2.38 bpw    rfn_error: 0.03793
 -- 0.1:4b/0.4:3b/0.5:2b 32g s4    2.72 bpw    rfn_error: 0.03168
 -- 0.1:4b/0.9:3b 32g s4           3.22 bpw    rfn_error: 0.01941
 -- 1.0:3b 128g s4                 3.03 bpw    rfn_error: 0.02422
 -- 1.0:3b 32g s4                  3.13 bpw    rfn_error: 0.02196
 -- 0.05:4b/0.95:3b 32g s4         3.17 bpw    rfn_error: 0.02002
 -- 0.4:4b/0.6:3b 32g s4           3.53 bpw    rfn_error: 0.01706
 -- 0.6:4b/0.4:3b 64g s4           3.66 bpw    rfn_error: 0.01630
 -- 1.0:4b 128g s4                 4.03 bpw    rfn_error: 0.01259
 -- 1.0:4b 32g s4                  4.13 bpw    rfn_error: 0.01116
 -- 1.0:5b 128g s4                 5.03 bpw    rfn_error: 0.00671
 -- 0.1:6b/0.9:5b 32g s4           5.22 bpw    rfn_error: 0.00626
 -- 0.4:6b/0.6:5b 32g s4           5.53 bpw    rfn_error: 0.00590
 -- 1.0:6b 128g s4                 6.03 bpw    rfn_error: 0.00430
 -- 0.1:8b/0.9:6b 128g s4          6.23 bpw    rfn_error: 0.00397
 -- 1.0:8b 32g s4                  8.13 bpw    rfn_error: 0.00455
 -- Time: 742.49 seconds
 -- Layer: model.norm (RMSNorm)
 -- Layer: lm_head (Linear)
 -- Calibration perplexity (base): 5.3018
 -- Optimizing...
 -- rfn_error: 0.27881  bpw: 2.17262
 -- rfn_error: 0.13940  bpw: 2.84866
 -- rfn_error: 0.06970  bpw: 3.74662
 -- rfn_error: 0.03485  bpw: 4.77840
 -- rfn_error: 0.05228  bpw: 4.27941
 -- rfn_error: 0.06099  bpw: 3.90092
 -- rfn_error: 0.05663  bpw: 4.02832
 -- rfn_error: 0.05881  bpw: 3.96083
 -- rfn_error: 0.05772  bpw: 3.99127
 -- rfn_error: 0.05718  bpw: 4.00975
 -- rfn_error: 0.05745  bpw: 3.99444
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- Tokenizing samples...
 -- Token embeddings again...
 -- Quantizing...
 -- Layer: model.layers.0 (Attention)
Traceback (most recent call last):
  File "/home/thireus/exllamav2/convert.py", line 195, in <module>
    quant(job, save_job, model)
  File "/home/thireus/exllamav2/conversion/quantize.py", line 462, in quant
    outputs = module.forward(x, cache, attn_mask, intermediates = True)
  File "/home/thireus/exllamav2/exllamav2/attn.py", line 195, in forward
    return self.forward_torch(hidden_states, cache, attn_mask, past_len, intermediates)
  File "/home/thireus/exllamav2/exllamav2/attn.py", line 460, in forward_torch
    key_states = self.repeat_kv(key_states, self.model.config.num_key_value_groups)
  File "/home/thireus/exllamav2/exllamav2/attn.py", line 188, in repeat_kv
    hidden_states = hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
RuntimeError: CUDA error: invalid configuration argument
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

**Comment by turboderp** [comment]
It looks like you've run out of memory when moving from the smaller sample of the dataset used for measurement to the full sample used for quantization. I'm working on memory optimizations that should help with that. Expect an update in a little bit.

In the meantime, it seems you've successfully run the measurement pass, so the progress should be saved.

**Comment by Thireus** [comment]
I see! Indeed the progress is now saved and completed Measuring Quantization (-ml optional parameter was set to default 2048).

I was thinking that now the issue was with the -l parameter which I had set to 4096. So I edited the job.json and set it back to 2048, but that didn't help, still getting the same error.

When you say memory, do you mean RAM or VRAM? I currently sit with 6GB+128GBswap of free memory and 24GB of VRAM. When I run the job again, it takes 3-4 sec to produce the error, the RAM receives about 2GB and VRAM about 1.5GB before the error happens, so barely filling anything.

I'll try your updated code as soon as it drops, thank you so much for the support!

**Comment by turboderp** [comment]
The issue is definitely VRAM in that case. If you're hacking the job.json file you have to be very mindful of what data it's expecting to be in sync with the job progress. In your case it's already created the embeddings from the calibration data, at a row length of 4096, which it reads from input_states.safetensors.

If you change the progress to "optimize", it should jump back to the optimization step and recreate the embeddings with the updated settings.

Alternatively, since you're at this point, you've got the "measurement.json" file in the output folder. If you copy that to a different folder and delete the rest of the work folder, you can start a new conversion job specifying `-m <measurement_file>` and it should skip the measurement pass.

**Comment by Thireus** [comment]
Ah, thank you for the tip! Your suggestion worked.

Edited job.json with:
```
    "length": 2048,
...
    "progress": "optimize",
```

Job resumed successfully:
```
$ CUDA_DEVICE_ORDER=PCI_BUS_ID CUDA_VISIBLE_DEVICES=1 python convert.py -i ~/safetensor/model -o ~/EXL2/model_4bit -c ~/EXL2/0000.parquet -b 4.0 -hb 6
 -- Resuming job
 !! Note: Overriding options with settings from existing job
 -- Input: /home/user/safetensor/model
 -- Output: /home/user/EXL2/model_4bit
 -- Calibration dataset: /home/user/EXL2/0000.parquet, 100 / 16 (16) rows, 2048 tokens per sample
 -- Target bits per weight: 4.0 (decoder), 6 (head)
 -- Optimizing...
 -- rfn_error: 0.27881  bpw: 2.17262
 -- rfn_error: 0.13940  bpw: 2.84866
 -- rfn_error: 0.06970  bpw: 3.74662
 -- rfn_error: 0.03485  bpw: 4.77840
 -- rfn_error: 0.05228  bpw: 4.27941
 -- rfn_error: 0.06099  bpw: 3.90092
 -- rfn_error: 0.05663  bpw: 4.02832
 -- rfn_error: 0.05881  bpw: 3.96083
 -- rfn_error: 0.05772  bpw: 3.99127
 -- rfn_error: 0.05718  bpw: 4.00975
 -- rfn_error: 0.05745  bpw: 3.99444
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- rfn_error: 0.05731  bpw: 3.99950
 -- Tokenizing samples...
 -- Token embeddings again...
 -- Quantizing...
 -- Layer: model.layers.0 (Attention)
 -- Linear: model.layers.0.self_attn.q_proj -> 0.05:3b/0.95:2b 32g s4, 2.17 bpw
 -- Linear: model.layers.0.self_attn.k_proj -> 0.05:3b/0.95:2b 32g s4, 2.17 bpw
 -- Linear: model.layers.0.self_attn.v_proj -> 0.05:3b/0.95:2b 32g s4, 2.17 bpw
 -- Linear: model.layers.0.self_attn.o_proj -> 0.05:3b/0.95:2b 32g s4, 2.17 bpw
```

Thank you!

With your upcoming memory optimisation do you believe the measurement and quantization pass can be completed using 24GB of VRAM for -ml 4096 -l 4096? I remember you said leaving them to 2048 shouldn't matter much in the end, but I'm still just curious to see how much of an impact it has on the final model - and I believe even the tiniest improvements could be beneficial (as I understand there would be no impact on the final quantized model size).

**Comment by venxzw** [comment]
@Thireus Hello, considering your successful implementation on this project, may I ask you a few questions? Firstly, the required dataset for quantification has not been provided with samples. Can you provide a format; Secondly, have you attempted to overgeneralize other models besides llama?

**Comment by Thireus** [comment]
@venxzw, I have only played with **WizardLM 70B V1.0**. Quanitzation notes, parameters and scripts I've used and can recommend are all available here: https://huggingface.co/Thireus/WizardLM-70B-V1.0-FP32-5.0bpw-h6-exl2

**Comment by venxzw** [comment]
> @venxzw, I have only played with **WizardLM 70B V1.0**. Quanitzation notes, parameters and scripts I've used and can recommend are all available here: https://huggingface.co/Thireus/WizardLM-70B-V1.0-FP32-5.0bpw-h6-exl2

thank you very much.But I have one more question: is the dataset used in the quantization process in the format you recommend as shown below?
A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions.
USER:
{prompt}
ASSISTANT:

**Comment by Thireus** [comment]
Hi @venxzw, I haven't touched the calibration dataset, I've used wikitext2: https://huggingface.co/datasets/wikitext/tree/refs%2Fconvert%2Fparquet/wikitext-2-raw-v1/train. It's an interesting question you ask, I do not know what would be the implications of using a formatted dataset instead of a generic one.

**Comment by venxzw** [comment]
> Hi @venxzw, I haven't touched the calibration dataset, I've used wikitext2: https://huggingface.co/datasets/wikitext/tree/refs%2Fconvert%2Fparquet/wikitext-2-raw-v1/train. It's an interesting question you ask, I do not know what would be the implications of using a formatted dataset instead of a generic one.

Thank you very much for your answer. Based on how you quantified the wizard using this framework, I am aware that it should support other models, but I am not sure which files need to be modified. I used Baichuan at the time and then looked at the py files under this framework, and found that there are many things that need to be modified.

**Comment by venxzw** [comment]
@Thireus Thank you very much for your answer. Based on how you quantified the wizard using this framework, I am aware that it should support other models, but I am not sure which files need to be modified. I used Baichuan at the time and then looked at the py files under this framework, and found that there are many things that need to be modified.
